### DISCUSSION TITLE
How to use optimized mesh

### DISCUSSION THREAD

**Discussion by maxxnino** [zeux/meshoptimizer#410] [Q&A]
Sorry, I'm new to the library, and graphic programming in general. I use `gltfpack` to optimize gltf file. And using `cgltf` load the optimized gltf file. But the component type jump from `float32/uint16/int8` all over the place. How can I load and use it in `vulkan `renderer. What the format for each of them ( vertex, index, normal, tangent).

**Comment by Adi-Amazon** [comment]
Hi @maxxnino

Not sure I fully understand the question, I'll try to explain hoping it's not too much below your level.

There is no absolute answer to this - each component would have its own data stream and properties. For examples, vertices (position buffer/stream) will most likely be 3 x float32, and so will be normals, tangent and bi-tangent (the last two might also additional component).
Triangle indices will be held usually using uint16 or uint32 depending again on the definition of the model and the overall amount of vertices (<64K can be represented by 2 bytes).

You can go over meshoptimizer.h and review many of the declared functions and how to use them or follow the main.cpp Process function and dive into the functions as they go over the model.
The function parseObj is a minimal example how to read the '.obj' data of a mesh file.
To use the processed data from the library you can use dumpObj and pipe it into an '.obj' file that you can now visually test with '.obj' visualizer, or alternatively you can create your own obj serializer function if you are feeding more than the minimal data used here (positions, normals, UV).

I'd recommend downloading an open source visualizer to understand how this data is plugged into the renderer - basically using streams / buffers and their descriptors so that DX12 / Vulkan would associate them properly on the GPU.

**Reply by maxxnino** [reply]
Thank you for clarify. I know the issue is because of I use one function to load both object with mesh quantization on/off. so the position/normal switch back and forth from f32 to uint16...

**Reply by Adi-Amazon** [reply]
Ok, thanks for clarifying.
I see your point and based on what I see in the source files you must be right but since this is pretty much the purpose of this (optimization of the mesh, compression and changes so allow better utilization by the GPU) it might be expected.
Going over the README.md can give you more details and they do mention that the format requires using specific loaders that will support it - here is one of the entries in the Doc: 'gltfpack substantially changes the glTF data by optimizing the meshes for vertex fetch and transform cache'.
I suggest going over this document to see if it can help you.

Quick question, I used the library for optimizing and creating meshlets given a mesh and using almost any '.obj' reader or reading the data from the generated file worked for me.  Is there a reason why you want to jump to the more 'exotic' optimization and compression?  While more optimized, this might not be the right first step for a POC and I would skip that until you have a stable read and render of the files.

**Reply by maxxnino** [reply]
I just use the default option `gltfpack` give me. I read the readme.md a couple times. But I still don't understand the mesh quantization option. Maybe I should leave the option off for now
```
gltfpack -i mesh.gltf -o mesh_fast.gltf
```

**Comment by maxxnino** [comment]
I have one more question, maybe outside the scoop of this library. But how can I use uint16 as position/text_coord and int8 as normal/tangent inside of glsl shader. It's only have vector of float

**Reply by Adi-Amazon** [reply]
This might give you a short intro on this: https://www.khronos.org/opengl/wiki/Normalized_Integer

While requires more understanding on the compression side of the algorithm, on the shader side the GPU / shader compiler will do that for you, and you can try and verify.

**Reply by maxxnino** [reply]
I'm using vulkan. so the vertex shader I keep the same. And only change the vertex input attribute format for `position` from `VK_FORMAT_R32G32B32_SFLOAT`  to `VK_FORMAT_R16G16B16_UNORM`. But it's doesn't help. It mess up the mesh so much. I think it not simple as a switch like that
```glsl
layout(location = 0) in vec3 in_position;
layout(location = 1) in vec2 in_tex_coord;
layout(location = 2) in vec3 in_normal;
layout(location = 3) in vec4 in_tangent;

```

**Reply by Adi-Amazon** [reply]
I suggest you ping @zeux directly with the question how to match the expected format. Depending on the compression, the format you need to pass to the shader can be very different. The sources for both compression and decompression are provided, so maybe you can do a deep dive on that.
An alternative that I used on my side was to create the library and use it to optimize and generate meshlets. Having done that I can now both dump it as a 3D file myself (.obj is very simple to use for that - dumpObj is provided but you can extend it) and/or load to my own data structures / buffers for rendering. Perhaps this is an approach you can try.

**Reply by maxxnino** [reply]
I manually add a plane, and texcoord, with uint16 format, and it work find. The vulkan api auto convert to vec3 on the fly. I thing the problem is cgltf don't support quantization extension. Or I need decode the quantization value back

**Reply by maxxnino** [reply]
I know the problem is. For position the type is vec3 but stride is 8 bytes not 6 bytes.
